metasrv/src/watcher/watcher_manager.rs

```diff
@@ -0,0 +1,276 @@
+//  Copyright 2021 Datafuse Labs.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+use core::ops::Range;
+use std::collections::BTreeMap;
+use std::sync::Arc;
+
+use common_base::tokio;
+use common_base::tokio::sync::mpsc;
+use common_base::tokio::sync::mpsc::Sender;
+use common_meta_raft_store::state_machine::StateMachineSubscriber;
+use common_meta_types::protobuf::event::EventType;
+use common_meta_types::protobuf::watch_request::FilterType;
+use common_meta_types::protobuf::Event;
+use common_meta_types::protobuf::WatchRequest;
+use common_meta_types::protobuf::WatchResponse;
+use common_meta_types::SeqV;
+use common_range_set::RangeSet;
+use common_tracing::tracing;
+use tonic::Request;
+use tonic::Status;
+
+use super::WatcherKey;
+use super::WatcherStream;
+
+pub type WatcherId = i64;
+pub type WatcherStreamSender = Sender<Result<WatchResponse, Status>>;
+pub type CloseWatcherStreamReq = (WatcherId, String);
+
+type CreateWatcherEvent = (Request<WatchRequest>, WatcherStreamSender);
+
+#[derive(Clone, Debug)]
+struct StateMachineKvData {
+    pub key: String,
+    pub event: EventType,
+    pub prev: Option<SeqV>,
+    pub current: Option<SeqV>,
+}
+
+#[derive(Clone, Debug)]
+pub struct WatcherStateMachineSubscriber {
+    sm_tx: mpsc::UnboundedSender<StateMachineKvData>,
+}
+
+#[derive(Debug)]
+pub struct WatcherManager {
+    /// A channel for sending create watch request.
+    create_tx: mpsc::UnboundedSender<CreateWatcherEvent>,
+
+    shutdown_tx: mpsc::UnboundedSender<()>,
+
+    pub subscriber: WatcherStateMachineSubscriber,
+}
+
+#[derive(Debug)]
+struct WatcherManagerCore {
+    /// A channel for receiving create watcher request from grpc service.
+    create_rx: mpsc::UnboundedReceiver<CreateWatcherEvent>,
+
+    sm_rx: mpsc::UnboundedReceiver<StateMachineKvData>,
+
+    /// A channel for sending stop stream request.
+    close_stream_tx: Arc<mpsc::UnboundedSender<CloseWatcherStreamReq>>,
+
+    /// A channel for receiving stop stream request.
+    close_stream_rx: mpsc::UnboundedReceiver<CloseWatcherStreamReq>,
+
+    shutdown_rx: mpsc::UnboundedReceiver<()>,
```

Comment:
What about merging all `*_rx` into one mpsc channel?

Such an `rx` can be used to receive every kind of event.
Of cause it requires defining the event with an `enum`.
Managing several enum variants may be easier than managing several channels.

And when this `rx` is shutting down, `WatchManager` will be shut down.

It is similar to `close_stream_tx`: a `closing` event can be replaced with shutting down a channel in most cases.

Reply:
I do not see why you always click `resolve` before making any update...